### PR TITLE
Add Image classification

### LINKED ISSUE #446: Add classification tool
### Tasks
(will eventually be split in several issues at grooming)

- [x] Add the tag tool to the list of tools in the labelling interface
- [x] When a class is selected with the tag tool selected, a tag of the given class is added (tags are using Chakra UI component: Tag/Square/Icon Left/Small)
- [x] There cannot be two tags of the same class
- [x] A tag can be removed by clicking on its `x`
- [x] The class name inside the tag is either white or chakra gray-800 whichever maximizes the contrast
- [x] The class selection menu width needs to be set to a maximum to avoid pushing the tags all the way to the right (today a very long class name occupies the complete screen width)
- [x] Right click on the image displays the context class selection menu and also allows to add tag
- [x] Add `tag` as a label type
- [x] ~~Update COCO export resolver to export tags~~ Moved to https://github.com/labelflow/labelflow/issues/485
- [x] ~~Update YOLO export resolver to export tags~~ Moved to https://github.com/labelflow/labelflow/issues/485

### Questions / Remarks for discussion
- In terms of UX/UI it would be better when the class selection menu is expanded to be able to select/unselect multiple class.
- When we will have tools with options (accuracy of smart tools for instance), it will move tags further to the right
- Set a maximum width for tags (in the case a class name is very long)
- There can be several rows of tags if there are many many?
- Tags interfere with the drawing area, to test in practice if its problematic

### Design
[Mockup (source of truth)](https://www.figma.com/file/4J9xyfF5v4B98MuKVO0Y8V/LabelFlow-mockups?node-id=2728%3A15203)
image

### PR DESCRIPTION
**Pull request by crubier**
# Feature

## Work performed

Closes #446 

- [x] Add classification tool
- [x] Add [K] shortcut to this tool
- [x] Add [C] shortcut to open context menu consistently for the classifications just like it was already for the boxes and polygons
- [x] Add [D] shortcut to deselect selected label (it allows to label classifications with keyboard only)
- [x] Created a "Classification tag" component
- [x] Created a "Classification overlay in openlayers" component
- [x] Added these two components to the app
- [x] Add GraphQL enum value for LabelType.Classification 
- [x] Add mutations to create classifications and "switch" them on/off
- [x] Refactor the effects a bit to be able to reuse them for classification creation
- [x] Optimistic response ? cache ? to avoid flashing the classifications when adding/removing one
- [x] Change the behaviour of the class selection option menu when in classification mode to show "Pick a class to apply"
- [x] Added stories of these components
- [x] Fixed buggy behaviour 1 (https://github.com/labelflow/labelflow/pull/447#issuecomment-925290696)
- [x] Fixed buggy behaviour 2, by adding the `key` property to the `<olInteractionPointer
          key={selectedTool}>` in `labelflow/typescript/web/src/components/labelling-tool/openlayers-map/select-and-modify-feature/select-interaction.tsx`  (https://github.com/labelflow/labelflow/pull/447#issuecomment-927692252)
- [x] Add a "createLabelClassAndCreateLabel" effect for when you create a label type when in classification mode to solve buggy behaviour 3 (https://github.com/labelflow/labelflow/pull/447#issuecomment-927785513)
- [x] When selecting an existing classification label, if you change it's class, ensure that it does not create a duplicate classification.
- [x] Fix unit tests post refactor
- [x] Add these corner case above to e2e tests
- [x] Add integration tests of classification

### CODE REVIEW

**Comment by nicOwlas** [comment]
I would change the label icon for[ this one ](https://www.figma.com/file/4J9xyfF5v4B98MuKVO0Y8V/LabelFlow-mockups?node-id=2725%3A11710)(the present one looks like a polygon at first glance)
image

**Comment by crubier** [comment]
> I would change the label icon for[ this one ](https://www.figma.com/file/4J9xyfF5v4B98MuKVO0Y8V/LabelFlow-mockups?node-id=2725%3A11710)(the present one looks like a polygon at first glance)
> image

Try to find one on react-icons

**Comment by nicOwlas** [comment]
> Try to find one on react-icons
BsTag
image

**Comment by nicOwlas** [comment]
The class context menu gets in the way when creating the first point of a bounding box (like if we were doing a right click on an existing label)
image

**Comment by nicOwlas** [comment]
For the icon, could we change it one more time? Sorry, the present one I advice you to take has a thinner stroke than the bounding box and polygon icons. `BiPurchaseTagAlt` is better.
image

**Comment by crubier** [comment]
> Did a new version deploy @crubier ? I still see the same bug with the class context menu appearing on left click.
> 
> Also the class selection menu remains on `None`, no matter the tag I select. It is not coherent with the behaviour we have when drawing bounding boxes or polygons for which the class selection menu displays the selected class.

No indeed the redeploy had failed, relaunched it

**Comment by nicOwlas** [comment]
> This is where mockups are useful (I tried the icon you wanted and thought it was too thin too, but at this point I just want to let you decide). But I know the classification was not planned so fine

On Figma I had thicken the stroke to make it equal to the other icons, but yeah I understand that's not something straight forward when importing a react icon.

**Comment by crubier** [comment]
> > This is where mockups are useful (I tried the icon you wanted and thought it was too thin too, but at this point I just want to let you decide). But I know the classification was not planned so fine
> 
> On Figma I had thicken the stroke to make it equal to the other icons, but yeah I understand that's not something straight forward when importing a react icon.

Then in that case if you don't want a developer to spend 30min on it, ideally you'd need to export the Figma icon as a SVG and include it in the Issue. (And even doing so is quite cumbersome because it still has to be imported differently than other icons in the app). Really the best solution is to just pick a react-icon as much as possible

**Comment by crubier** [comment]
Modified the class selection menu when in classification mode to get better UX:

Screenshot 2021-09-21 at 23 08 05
>

**Comment by nicOwlas** [comment]
Good to have converted the text in the class selection to menu to `+ Add a label`  👍

Strange behaviour for the rest of my test:

**On first try:**
everything worked well (but right click is deactivated in classification mode).

**On the second try (after refresh) buggy behaviour:**
Looks like I can add only one class. And in bounding box mode, when I right click on a label and select a class it adds a tag instead of changing the class of the nearest bounding box.
Recording: 
https://user-images.githubusercontent.com/16015322/134313499-55925c3b-1cdd-409e-a6c8-d2ef94c34a0b.mov

**On third try**
It worked well again (but right click in classification mode)

**Comment by nicOwlas** [comment]
The debug info on the version that is bugging

```
LabelFlow
Debug





Debug information
Links:
Link to Github issue tracker
Link to Sentry error tracker
Link to Clarity usage analytics
Link to client bundle analysis
Link to server bundle analysis
Link to Graphiql
Link to Database admin
Link to CMS admin
Link to Github branch
Link to Github commit
Link to Vercel deployment
Link to CMS deployment on Heroku
Link to mailer provider
Client environment:
{
  "serverType": "Standard LabelFlow Client App",
  "isInWindowScope": true,
  "isInServiceWorkerScope": false,
  "isDevelopmentEnvironment": false,
  "browser": {
    "name": "chrome",
    "version": "93.0.4577",
    "os": "Mac OS",
    "type": "browser"
  },
  "env": {
    "NODE_ENV": "production",
    "NEXT_PUBLIC_VERCEL_ENV": "preview",
    "NEXT_PUBLIC_VERCEL_URL": "labelflow-rl4w9ocej-labelflow.vercel.app",
    "NEXT_PUBLIC_VERCEL_GIT_PROVIDER": "github",
    "NEXT_PUBLIC_VERCEL_GIT_REPO_SLUG": "labelflow",
    "NEXT_PUBLIC_VERCEL_GIT_REPO_OWNER": "labelflow",
    "NEXT_PUBLIC_VERCEL_GIT_REPO_ID": "",
    "NEXT_PUBLIC_VERCEL_GIT_COMMIT_REF": "feature/classification",
    "NEXT_PUBLIC_VERCEL_GIT_COMMIT_SHA": "44459c0e2315520303478c1806a0d63c2f03c6e8",
    "NEXT_PUBLIC_VERCEL_GIT_COMMIT_MESSAGE": "Fix lint",
    "NEXT_PUBLIC_VERCEL_GIT_COMMIT_AUTHOR_LOGIN": "crubier",
    "NEXT_PUBLIC_VERCEL_GIT_COMMIT_AUTHOR_NAME": "crubier"
  }
}

Server environment:
{
  "serverType": "Standard LabelFlow Service Worker Server",
  "isInWindowScope": false,
  "isInServiceWorkerScope": true,
  "isDevelopmentEnvironment": false,
  "browser": {
    "name": "chrome",
    "version": "93.0.4577",
    "os": "Mac OS",
    "type": "browser"
  },
  "env": {
    "NODE_ENV": "production",
    "NEXT_PUBLIC_VERCEL_ENV": "preview",
    "NEXT_PUBLIC_VERCEL_URL": "labelflow-rl4w9ocej-labelflow.vercel.app",
    "NEXT_PUBLIC_VERCEL_GIT_PROVIDER": "github",
    "NEXT_PUBLIC_VERCEL_GIT_REPO_SLUG": "labelflow",
    "NEXT_PUBLIC_VERCEL_GIT_REPO_OWNER": "labelflow",
    "NEXT_PUBLIC_VERCEL_GIT_REPO_ID": "",
    "NEXT_PUBLIC_VERCEL_GIT_COMMIT_REF": "feature/classification",
    "NEXT_PUBLIC_VERCEL_GIT_COMMIT_SHA": "44459c0e2315520303478c1806a0d63c2f03c6e8",
    "NEXT_PUBLIC_VERCEL_GIT_COMMIT_MESSAGE": "Fix lint",
    "NEXT_PUBLIC_VERCEL_GIT_COMMIT_AUTHOR_LOGIN": "crubier",
    "NEXT_PUBLIC_VERCEL_GIT_COMMIT_AUTHOR_NAME": "crubier"
  }
}

LabelFlow | Image tutorial-image-1
```

**Comment by crubier** [comment]
I think this strange behaviour is due to caching of the code, the third try is probably what I've coded, the previous tries are bugs when the cache / service worker is being updated. I think adding Sentry fucked up something around this

**Comment by nicOwlas** [comment]
Some recordings where I'm able to change some bounding box class while in classification mode and add 2 tags of the same class when switching between bounding box tool and classification tool. In classification mode, right click will sometimes change the class of the nearest bounding box and sometimes add/remove a class.

Uploading Screen Recording 2021-09-22 at 18.52.17.mov…

I also add another buggy behaviour:

1. I right click on an image in classification mode
2. I select a class in the menu
3. The name of the class is copied into the search/add bar
4. The label class is not actually changed.

**Comment by crubier** [comment]
@nicOwlas you need to give cleared reproductions, I don't understand your scenarios, and don't see the clicks / key press on the videos so it's impossible to understand anything. 

Please make a precise list of steps of how to reproduct problematic behaviours

**Comment by nicOwlas** [comment]
It's not really clear to me too how to exactly reproduce but creating and deleting labels/tags and switching between both modes triggers it.

**Comment by nicOwlas** [comment]
Could you confirm what sha I'm suppose to read in the debug info? There is very often the update modal which appears so I'm never sure if I'm actually on the latest.

**Comment by crubier** [comment]
> It's not really clear to me too how to exactly reproduce it but creating and deleting labels/tags and switching between both modes triggers it.

I’ll try to put that in the Test function, we’ll see if it works. I really need a reproduction for this kind of thing, we can’t work without.

**Comment by crubier** [comment]
> Could you confirm what sha I'm suppose to read in the debug info? There is very often the update modal which appears so I'm never sure if I'm actually on the latest.



> Could you confirm what sha I'm suppose to read in the debug info? There is very often the update modal which appears so I'm never sure if I'm actually on the latest.

Check the commit list of this PR and compare the SHA

**Comment by nicOwlas** [comment]
I checked the sha if the last 10 commits. It corresponds to none. To what does it correspond to the debug info sha?
`"NEXT_PUBLIC_VERCEL_GIT_COMMIT_SHA": "01f2b06bae90b30d9404b4e353c8fa68bda02bc0"`

**Comment by crubier** [comment]
> I checked the sha of the last 10 commits. It corresponds to none. To what does it correspond to the debug info sha?
> `"NEXT_PUBLIC_VERCEL_GIT_COMMIT_SHA": "01f2b06bae90b30d9404b4e353c8fa68bda02bc0"`
> 
> Edit, I checked using the github search and this sha does not corresponds to any commit.



> I checked the sha of the last 10 commits. It corresponds to none. To what does it correspond to the debug info sha?
> `"NEXT_PUBLIC_VERCEL_GIT_COMMIT_SHA": "01f2b06bae90b30d9404b4e353c8fa68bda02bc0"`
> 
> Edit, I checked using the github search and this sha does not corresponds to any commit.

Indeed, it looks the vercel git commit sha does not match the sha from GitHub...  strange. But the commit message matches the last commit so I guess it’s fine

**Comment by nicOwlas** [comment]
Ok so I cleared everything and started from a clean sheet. Lets fix them one by one to avoid some bad coupling.

First one:
1. On tutorial image 1: select the bounding box tool
2. Draw a bounding box around the drone with the class none
3. Go to the class selection menu and select the class `Horse`
4. **Error**: It does not get selected. The menu displays none only.
5. Right click on on the bounding box and select `Horse`: it works.

Recording. I only did some left clicks except in then end, one right click to display the class selection menu on top of the bounding box.

https://user-images.githubusercontent.com/16015322/134414259-dd37d305-a6c8-439a-bc93-18905e5debfa.mov

**Comment by crubier** [comment]
@nicOwlas this bug is now fixed. 

I found a more subtle one now:

- On tutorial image 1: select the bounding box tool
- Draw a bounding box around the drone with the class none
- Go to the class selection menu and select the class Horse
- FIXED: It does get selected. The menu displays Horse (of course the bounding box does not change color)
- Right click on on the bounding box and select Horse: it works.
- Press "k" to move to classification tool
- Right Click outside of the previous box to open the context menu to add a label
- ERROR: It does not open

**Comment by crubier** [comment]
@nicOwlas second bug fixed now.

Now new buggy behaviour that I had identified already:

- In classification mode
- Click Add a Label button to add a label
- Enter a name for a new Label
- Type enter
- The new class got created, but did not get applied.

**Comment by crubier** [comment]
@nicOwlas I need some QA done on this branch. Can you have a look a try stuff ? There is at least one bug I know exist but I wont tell you, so you search thoroughly

**Comment by nicOwlas** [comment]
> @nicOwlas I need some QA done on this branch. Can you have a look a try stuff ? There is at least one bug I know exist but I wont tell you, so you search thoroughly

Is this a joke? Next time I find a bug I will also won't tell you how to reproduce it so that you search thoroughly too.

**Comment by crubier** [comment]
> > @nicOwlas I need some QA done on this branch. Can you have a look a try stuff ? There is at least one bug I know exist but I wont tell you, so you search thoroughly
> 
> Is this a joke? Next time I find a bug I will also won't tell you how to reproduce it so that you search thoroughly too.

No it’s not a joke at all, it’s to ensure independent testing. Basic QA technique 

This feature needs to be tested a lot because there are at least 96 different cases (add class / remove class / create class / class already exists / classification mode and select a detection/ etc...). I tried a lot of them but I’m sure i didn’t even cover half of possible cases so I need someone else to try and cover as much as possible.

**Comment by nicOwlas** [comment]
Let's solve the bugs one by one. The one you found may have some ripple effect. Like last week. Once it's solved I'll do the QA.

**Comment by crubier** [comment]
Ok @nicOwlas I think I'm done functionally, there is no known bug, it's deployed so you can try it and try to find problems or remarks.

**Comment by nicOwlas** [comment]
- ✅ Merging labels
- ✅ Adding, Removing, editing labels
- ✅ Switching between bb, polygons
- ✅ Exports to COCO and YOLO

### Brute force - Adding multiple labels of the same class
- ❌ On firefox, I'm able to add multiple times the same label by quickly pressing its shortcut. When doing this it slows down the page by a lot and 
image
image

- ❌ On Brave I could reproduce the same but it was more difficult
image


- ❌ On Firefox, this also created multiple times the same class (with the same shortcut...!). Could not reproduce on Brave.
image

## Not related to classification but observed during those tests (will be opened in a separate issue)
- When zooming (pinching) on the navbar, a button or the gallery, the complete interface is zoomed but it does not appear in the settings of the browser. I was blocked several times because of this before understanding how to undo this. Refreshing does not reset the zoom level.
image


## Things observed once but that I could not reproduce a second time

### Label not added
- Click on `+ Add a label` menu
- Type a new name
- The class is added but not the label

### Shortcut not working
- Select the bounding box tool
- The shortcuts `v`, `p`, `k` to switch back to another tool do not work

**Comment by crubier** [comment]
- When in mode classification, can't edit box or polygons

File: typescript/web/src/components/labelling-tool/openlayers-map/select-and-modify-feature/select-interaction.tsx [hunk 1]
```diff
@@ -34,6 +40,15 @@ export const SelectInteraction = ({
   useHotkeys(
     keymap.openLabelClassSelectionPopover.key,
     () => {
+      if (selectedTool === Tools.CLASSIFICATION) {
+        setIsContextMenuOpen(true);
+        setContextMenuLocation([
+          (image?.width ?? 0) / 2,
+          (image?.height ?? 0) / 2,
+        ]);
+        return;
```
Comment (etiennedupont) [inline typescript/web/src/components/labelling-tool/openlayers-map/select-and-modify-feature/select-interaction.tsx:49]: A bit useless now, should use tag's context menu

**Comment by crubier** [comment]
Fixed all things we saw during review, merging
